### PR TITLE
Add Memory Bank and MoCo

### PR DESCRIPTION
**Pull request by philippmwirth**
# Add Memory Bank and MoCo

I have added a memory bank parent class for the NTXentLoss. This way the loss can be used with and without memory bank. Furthermore, I have implemented MoCo, addressed issues #7 and #8, and updated a few docstrings.

## Changes
- Added a memory bank module as parent class for the NTXentLoss function. The NTXentLoss will query the memory bank for negative samples. If the size of the memory bank is zero, the NTXentLoss will continue as before. Otherwise, it will use the negative samples from the memory bank to calculate the loss.
- Added an example to the NTXentLoss docstring.
- Added `memory_bank_size: 0` argument to the config file. This should not be changed at the moment as SimCLR tends to diverge.
- Added default collate functions for `SimCLR` and `MoCo` for user friendliness.
- Addressed issue #8 for efficiency reasons.
- Added the possibility to use `ImageCollateFunction` without Gaussian blur.
- Embeddings can now be trained with a learning rate scheduler. 
- Added an implementation of the MoCo architecture.

## Why?
- The way the NTXentLoss with memory bank is implemented allows a user to choose any type of embedding model and transformation and use them together with the NTXentLoss irrespective of whether a memory bank is required or not. 
For example, a user could try any of the following combinations:
   - SimCLR & no memory bank -> like the paper
   - SimCLR & memory bank -> diverges, unfortunately
   - MoCo & no memory bank -> ???
   - MoCo & memory bank -> like the paper
- This flexible design should be continued for all loss functions which are added in the future.

### CODE REVIEW

**Review by IgorSusmelj — APPROVED** [review]

**Comment by IgorSusmelj** [comment]
MOCO and memory bank seems not to work properly for me:
Blue is without memory bank. Orange with memory bank. Trained for 100 epochs on CIFAR10.

image

**Comment by IgorSusmelj** [comment]
Memory bank should not update when `requires_grad=False`
